dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/ZipGitHubFileTree.java

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright 2024 OICR and UCSC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package io.dockstore.webservice.helpers;
+
+import io.dockstore.common.Utilities;
+import io.dockstore.webservice.CustomWebApplicationException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.channels.SeekableByteChannel;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import org.apache.commons.compress.archivers.zip.ZipArchiveEntry;
+import org.apache.commons.compress.archivers.zip.ZipFile;
+import org.apache.commons.compress.utils.SeekableInMemoryByteChannel;
+import org.apache.commons.io.IOUtils;
+import org.apache.http.HttpStatus;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Abstracts the file tree corresponding to a reference to a GitHub repository.
```

Comment:
This starts to look a lot like https://commons.apache.org/proper/commons-vfs/commons-vfs2/apidocs/org/apache/commons/vfs2/provider/zip/ZipFileSystem.html

Reply:
Sure, vfs2 and the `FileTree` implementations do share a bit of similar functionality.  However, there's lot of good reasons to use our own `FileTree` interface:

* easy to modify, for example, to add support for binary files or querying file types.
* easy to add operations, for example another way to query the tree.
* we control the semantics.
* the operations are limited to what we need, making the code easier to understand and eliminating irrelevant methods which could be harmful if used.

Also, don't see GitHub support in vfs2, and even if there is, imagine using vfs2 to implement an equivalent to our `ZipGitHubFileTree`, with the fallbacks + our auth scheme, etc, is going to end up being just about as complicated as  (or maybe more so than) what we've got now.

So, for now, recommend we use our own FileTree implementations, they're pretty simple.  And, down the line, should we need to, we can use vfs2 to implement FileTrees (behind our interface, under the hood) that support more complicated schemes like a file tree rooted at a given url.

Reply:
> there's lot of good reasons to use our own FileTree interface

I see where you're coming from, but there are also advantages to using a pre-existing library, particularly one that we already use elsewhere on the classpath. By re-using code, that is a potential bolus of code that we do not need to maintain, test, or document. As a simple example, we needed to document this to a certain extent for the rest of the team to understand what the aim of this was whereas vfs has good documentation that we can just refer others to. There are pre-existing integrating libraries we can look into and it also opens up the possibility that we can contribute code and have others help maintain it for us. 

> Also, don't see GitHub support in vfs2, and even if there is, it's very possible that using vfs2 to implement an equivalent to our ZipGitHubFileTree

I'm not sure what makes `ZipGitHubFileTree` all that GitHub aside from it being initialized using `gitHubSourceCodeRepo`. Presumably that's something we could do by extending `ZipFileSystem` with a new constructor. 

More to your bullet points, since the code is open-source, we can do much of this. We can override and disable irrelevant methods, we can add operations to query the tree, and we can add support for binary files or querying file types.

Reply:
I also noticed https://github.com/mmounirou/commons-vfs-git also gives an example using jgit, though I wouldn't suggest using it directly without forking/modification, just as inspiration

Reply:
Here's more about the `FileTree` interface and its motivation.  Background: I've spent a lot of time working in/amongst the webservice's file access and registration code, and I've been pondering the subsequent refactor for years...

Currently, in our codebase, our `.dockstore.yml`-based entry registration logic code is tightly coupled to and interwoven with our GitHub repository access logic, thus limiting `.dockstore.yml`-based  registrations to GitHub repositories.  However, by splitting apart the intertwined concepts (the notion of a "file tree", the registration logic, and the repository access primitives) into separate objects, we could support other sources for entry registrations (such as registering from a different source code repo site [GitLab, whatever comes down the pike in a few years, etc], a Zip file, local directory, etc), improve testing by making it easier to mock the input files, etc.

To that end, this PR introduces `FileTree`, a simple interface that abstracts read-only access to a tree of files.  The `FileTree` interface consists of two methods (`readFile` and `listFiles`) that were essentially lifted from the existing `SourceCodeRepoInterface` and have the same behavior, and a new method `listPaths`, which enumerates the files in the tree (to avoid having to enumerate the file paths using the other methods, which can be very slow when there's lots of directories).  

## Good things about the `FileTree` interface

The `FileTree` interface is:

* Understandable: `FileTree` consists of three methods, two of which were essentially lifted from `SourceCodeRepoInterface`, have the same contract, and are already familiar to us.
* Lightweight: The three production-oriented `FileTree` interfaces/classes (`FileTree`, `CachingFileTree`, and `ZipGitHubFileTree`) are split into bite-sized implementations that are easy to read, understand, reason about, and test.  Number of non-comment/blank lines within the class body of each:
  - FileTree: 3 lines, 
  - CachingFileTree: 32 lines, 
  - ZipGitHubFileTree: 55 lines.
* Under our control: We define the semantics.  That is, exactly what the methods do, what the arguments mean, what the results are, and which exceptions get thrown and when.  We also control additions and modifications to the interface.
* Compatible: `FileTree` provides the operations that our code already uses, making it easy to change code to use it without lots of rewriting.
* Sufficient: `FileTree` provides the full set of necessary file access operations.
* Focused: The interface does not define functionality that we don't need and won't use, avoiding later confusion and accidental use.
* Consumable: The interface is defined such that code that uses the interface doesn't need to heavily process arguments or results, remap exceptions, etc.
* Extensible: We can easily wrap FileTrees in other FileTrees to implement additional behaviors.  For example, in about 20 lines of non-comment/blank/import code, to prevent huge downloads, we can create a wrapper FileTree implementation that tracks the total size of downloaded files and throws when it reaches a specified limit. 

Note that `FileTree` isn't intended to provide general-purpose access to a filesystem.  Rather, it's a compact abstraction, lifted out of an existing abstraction, that allows read-only access to a tree of files, providing a simple set of operations, the majority of which our code currently uses.  It fits well.


## Vfs2 overview

The vfs2 library abstracts general-purpose filesystem access.  The core interface is `FileSystem`, but caching is implemented via a different interface that doesn't implement it, which is strange.  Code quality is ok.  vfs2 is sparsely documented, with at least one `TODO` in the current documentation.  vfs2 does  not include GitHub support, although there appears to be some homespun 3rd party implementations out there.  

## Vfs2 as our core file tree access interface

It has been proposed that we replace `FileTree` with an abstraction from vfs2.

After deep consideration, I conclude that we should not use the vfs2 `FileSystem` interface as the main way that our code reads source files.  It does not have most of the good qualities of the `FileTree` interface (which I listed above).  As such, we'd have to hack it (write a substantial amount of code and perhaps maintain our own fork) to make vfs2 useful to us in this context.  And, after all that effort, the result might very well be kludgey and unreliable, and we'd have written a bunch of code that we'd need to maintain, which is what we were trying to avoid by using vfs2 in the first place...

## Vfs2 as the core of ZipGitHubFileTree 

It has been proposed that we use vfs2's ZipFileSystem implementation as the underpinning for ZipGitHubFileTree.  After deep consideration, I recommend against doing so.

To handle the file prefix that GitHub inserts into its Zipball, and special treatment for symlinks, ZipGitHubFileTree must manipulate the ZipBall on a low level (via apache's ZipFile), and it will be difficult to implement equivalent behavior by "riding on top" of the Zipball via vfs2's ZipFileSystem.  We'll also need code to handle fallbacks to the existing `GitHubSourceCodeRepo` implementation.  Additionally, low-level Zipball access allows us to later implement features such as avoiding potential DOSes by refusing to unzip very large Zip entries by first inspecting the included `size` field, before expansion.

Reply:
Given that the current PR contains a reasonable implementation that is already written, and that we only use `FileTree` in the `.dockstore.yml` inference code, I propose that we merge it.  Then, we can see how it performs, and depending on the results, decide to keep it, replace it, etc.  It has small footprint, appears to work well, and seems to have lots of good qualities, not sure what the rationale is for ripping it out now, without evidence that it's defective.  If some of us are convinced that vfs2 is the way to go, we can implement the vfs2 alternative and compare them.

Reply:
See thread at https://oicr.slack.com/archives/C05EZH3RVNY/p1722828684206789

Rather than address vfs2 point by point. It turns out there's an alternative to vfs2 built into the Java JDK with some interesting implementations and another zip filesystem as well out of the box.  https://docs.oracle.com/javase/8/docs/technotes/guides/io/fsp/zipfilesystemprovider.html

There are some interesting implementations of that interface for testing https://github.com/google/jimfs and for GitHub integration https://github.com/gnodet/githubfs that  we can take inspiration from. Even a familiar face at https://github.com/stain/nio-zipfs

There are probably other alternatives and implementations of those alternatives if we want to shop around more. 

>  It has a small footprint, appears to work well, and seems to have lots of good qualities, not sure what the rationale is for ripping it out now, without evidence that it's defective.

I think the perspective I'm coming from is more the other way around i.e. why should we be putting effort into a new filesystem abstraction when there are existing Java abstractions 

Via feature creep, a seemingly small implementation can grow until we kinda forget the reason why we decided to go down the route of "not invented here."  Remember, the intent is 

> We should refactor all of our code similarly.

i.e. broad and wide-ranging

> Create a synthetic in-memory implementation

i.e. sounds like something like Jimfs

I think it's worth the effort into seeing what is out there already and what we can re-use. We're not really experts in file systems and this isn't really the kind of thing that we can write into grants or easily justify effort into. So it reasons that we should put a bit more effort into shopping around for something maintained by others that we can re-use and maybe even contribute to if you have an interest.